src/plugins/discover/public/application/main/components/layout/discover_documents.tsx

```diff
@@ -75,49 +75,75 @@ function DiscoverDocumentsComponent({
   onAddFilter?: DocViewFilterFn;
   savedSearch: SavedSearch;
   setExpandedDoc: (doc?: DataTableRecord) => void;
-  state: AppState;
-  stateContainer: GetStateReturn;
+  stateContainer: DiscoverStateContainer;
   onFieldEdited?: () => void;
 }) {
   const { capabilities, dataViews, uiSettings } = useDiscoverServices();
+  const [query, sort, rowHeight, rowsPerPage, grid, columns, index] = useAppStateSelector(
+    (state) => {
+      return [
+        state.query,
+        state.sort,
+        state.rowHeight,
+        state.rowsPerPage,
+        state.grid,
+        state.columns,
+        state.index,
+      ];
+    }
+  );
+
   const useNewFieldsApi = useMemo(() => !uiSettings.get(SEARCH_FIELDS_FROM_SOURCE), [uiSettings]);
   const hideAnnouncements = useMemo(() => uiSettings.get(HIDE_ANNOUNCEMENTS), [uiSettings]);
   const isLegacy = useMemo(() => uiSettings.get(DOC_TABLE_LEGACY), [uiSettings]);
   const sampleSize = useMemo(() => uiSettings.get(SAMPLE_SIZE_SETTING), [uiSettings]);
 
   const documentState = useDataState(documents$);
-  const isLoading = documentState.fetchStatus === FetchStatus.LOADING;
-  const isPlainRecord = useMemo(
-    () => getRawRecordType(state.query) === RecordRawType.PLAIN,
-    [state.query]
-  );
+  const isDataLoading = documentState.fetchStatus === FetchStatus.LOADING;
+  // This is needed to prevent EuiDataGrid pushing onSort because the data view has been switched.
+  // 1. When switching the data view, the sorting in the URL is reset to the default sorting of the selected data view.
```

Comment:
Here's the logic what's happening with the state when a data view is switched:
https://github.com/elastic/kibana/blob/1a213bd7cea43caee7ac02f2490c357c27544923/src/plugins/discover/public/application/main/utils/get_switch_data_view_app_state.ts#L38-L58